### ISSUE TITLE
groupBy followed by concatMap is hanging on low cardinality groups

### ISSUE THREAD

**Issue by EugenZ** [reactor/reactor-core#3442]
In the example below we defined 4 groups with 86 elements each the result is hanging test.
If we change number from 86 to 85 it works as expected. Or if we add bigger prefetch size it will finish successfully.
What we noticed that for default value of prefetch size is used implementation of queue `SpscArrayQueue` but for specified value of prefetch or Integer.MAX_VALUE is used queue `SpscLinkedArrayQueue` which works without hanging. Maybe this is not root cause of that issue but would be nice to know how to avoid hanging and proper usage of groupBy and concatMap with default implementation.

## Expected Behavior

Finish successfully.

## Actual Behavior

Hanging infinitely.

## Steps to Reproduce

```java
import lombok.*;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

import java.util.AbstractMap;

import static java.lang.String.valueOf;

@SpringBootTest
class ReactorhangingApplicationTests {

    @Test
    void contextLoads() {
        final Flux<Integer> ids = Flux.range(1, 4);
        var grouped = ids.flatMap(id ->
                        Flux.range(1, 86).map(n -> ReactorTestEntity.builder() // 85 - OK, 86 or any bigger number - hanging
                                .customId(id)
                                .price(valueOf(n))
                                .shares(valueOf(id + n))
                                .build()))
                .groupBy(ReactorTestEntity::getCustomId); // Adding prefetch size higher or Integer.MAX_VALUE will fix the issue

        var transformed = grouped.map(entry -> new AbstractMap.SimpleEntry<>(entry.key(), entry.map(this::fun).collectList()));
        var result = transformed
                .doOnNext(System.out::println)
                .concatMap(e -> e.getValue().map(list -> {
                    System.out.println(list);
                    return list;
                }))
                .log();

        StepVerifier.create(result).expectNextCount(4).verifyComplete();
    }

    String fun(ReactorTestEntity entity) {
        return entity.toString();
    }
}

@Data
@Builder(toBuilder = true)
@ToString
class ReactorTestEntity {
    public Integer customId;
    public String shares;
    public String price;
}
```

## Your Environment

* Reactor version(s) used: 3.5.5
* Other relevant libraries versions (eg. `netty`, ...): netty
* JVM version (`java -version`): v17, v18
* OS and version (eg `uname -a`): MacOS Ventura 13, Windows 10

**Comment by OlegDokuka** [comment]
Duplicate of #3427

**Comment by OlegDokuka** [comment]
That is expected behaviour. Backpressure mechanism prevents new elements to be consumed until downstream is ready to consume more. Since the concatMap remains open until inner Publisher is closed, no new elements are requested from the upstream, therefore groupBy can not produce new groups, therefore it does not request new events as well, therefore the source remains uncompleted with more elements to produce.

**Comment by EugenZ** [comment]
@OlegDokuka I guess it should be described in API doc somehow since it is not clear from documentation that `concatMap` will be in hanging state if prefetch value is missed in `groupBy`.
Also question about prefetch size in default implementation `groupBy` and with prefetch value Integer.MAX_VALUE size is 256 but second case works first one doesn't...

PS: thanks a lot for quick responses.

**Comment by OlegDokuka** [comment]
> @OlegDokuka did you look at description there is only 4 groups with 86 elements each even multiplied number 344 elements (let's imagine groups) hardly can be considered as high cardinality (according to the high cardinality definition "high-cardinality columns that can take hundreds of thousands of distinct values")?

Of course. concatMap can subscribe only to a single group. FlatMap subscribes to 4 inner producers. FlatMap has inner request of 32 elements from each once and then (32 - 32 >> 2). groupBy request 256 elements once and then every inner requests more once it produces some amount of elements (prefetch is not consistent compares to the other operators and depends on how every inner progresses over time)

so what happened, flatMap requested 32 for each inner which can produce 86 elements. All of them produced that number, 32 * 4 = 128, nothing happens after that since we wait until downstream (groupBy consumes that number). Now, groupBy requests 256, so flatMap starts producing buffered values, once any inner has produced 24 values, the `flatMap` sends a dedicated request for 24 more to be produced to it, so after all we should have extra 4 requests for 24 elements for every inner (in total 32 * 4 + 24 * 4 = 224 out of 256 requested), if we have consistent prefetch from groupBy, then not deadlock would appear since groupBy could request another number of elements, but the problem is that progress depends on inner group progress, and since only a single group is subscribed, this group requests more, thus once 24 elements a produced by inner, another request for 24 comes from the groupBy to flatMap ( 256 + 24 = 280), which in the meanwhile has received 24 more elements from each inner source, so another 24 is produced by id 1 and 24 by id 2, so group for id 1 (requests another 24 elements) and then around that point the progress stops since flatMap consumes from each group by round-robin rule. so in total, we should see the expected 32 * 4 + 24 * 4 + 24 * 4 + 24 = 344 values produced where the majority is buffered in groupBy inner buffer or in the flatMap inners buffers. Since progress can not be completed by inners since they still has more elements, groupBy can not complete existing inner groups, therefore concatMap cannot consume next group

**Comment by EugenZ** [comment]
@OlegDokuka appreciated for detailed explanations!
